### PR TITLE
daemon: add new accessChecker implementations

### PR DESCRIPTION
**Pull request by jhenstridge**
This has been extracted from PR #9043, and builds on top of PR #10126.

This PR includes the four implementations of the new `accessChecker` interface, without the changes to actually plumb them into the REST server.  The four checkers are intended to represent the different levels of access used by the REST API:

1. `openAccess` allows unrestricted access on the main `/run/snapd.socket` socket.
2. `authenticatedAccess` allows access on `/run/snapd.socket` provided one of the following is true:
    * The user has provided a macaroon login token.
    * The user is root
    * The user is granted access by polkitd
3. `rootAccess` allows access on `/run/snapd.socket` if the user is root.
4. `snapAccess` allows access on `/run/snapd-snap.socket` (for use by `/v2/snapctl`).

The branch also updates the existing `Command.canAccess` method to use the factored out `checkPolkitAction` helper function that was part of #9043.

### CODE REVIEW

File: daemon/access_test.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+		c.Fail()
+		return accessForbidden
+	}
+
+	var ac accessChecker = authenticatedAccess{}
+
+	req := httptest.NewRequest("GET", "/", nil)
+	user := &auth.UserState{}
+
+	// authenticatedAccess denies access from snapd-snap.socket
```
Comment (pedronis) [inline daemon/access_test.go:71]: could you add a test where Socket is neither snapd-snap.socket nor snapd.socket

File: daemon/access.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+const (
+	accessOK accessResult = iota
+	accessUnauthorized
+	accessForbidden
+	accessCancelled
+)
+
+var (
+	polkitCheckAuthorization = polkit.CheckAuthorization
+	checkPolkitAction        = checkPolkitActionImpl
```
Comment (pedronis) [inline daemon/access.go:44]: nitpick: could you split this out as its own var, it's a callable defined here below, it's a not a function from somewhere else

File: daemon/access_test.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+	ucred.Socket = dirs.SnapdSocket
+	c.Check(ac.checkAccess(nil, ucred, nil), Equals, accessOK)
+
+	// Access forbidden without peer credentials.  This will need
+	// to be revisited if the API is ever exposed over TCP.
+	c.Check(ac.checkAccess(nil, nil, nil), Equals, accessForbidden)
+}
+
+func (s *accessSuite) TestAuthenticatedAccess(c *C) {
+	defer func() {
```
Comment (pedronis) [inline daemon/access_test.go:57]: I know that these test are in daemon and not daemon_test,  I don't want to insist that they be moved to the latter in particular. But for consistency could you introduce Mock* () (restore func()) for this mocking?

File: daemon/access_test.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+	// The root user is granted access without a macaroon
+	ucred = &ucrednet{Uid: 0, Pid: 100, Socket: dirs.SnapdSocket}
+	c.Check(ac.checkAccess(req, ucred, nil), Equals, accessOK)
+}
+
+func (s *accessSuite) TestAuthenticatedAccessPolkit(c *C) {
+	defer func() {
+		checkPolkitAction = checkPolkitActionImpl
+	}()
```
Comment (pedronis) [inline daemon/access_test.go:94]: same

File: daemon/access_test.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+		return accessOK
+	}
+	ucred = &ucrednet{Uid: 42, Pid: 100, Socket: dirs.SnapdSocket}
+	c.Check(ac.checkAccess(req, ucred, nil), Equals, accessOK)
+}
+
+func (s *accessSuite) TestCheckPolkitActionImpl(c *C) {
+	defer func() {
+		polkitCheckAuthorization = polkit.CheckAuthorization
+	}()
```
Comment (pedronis) [inline daemon/access_test.go:128]: same

**Review by pedronis — COMMENTED** [review]
looks good, some comments and nitpicks

File: daemon/access_test.go (hunk 3 above)
Comment (jhenstridge) [reply daemon/access_test.go]: Done.  It turned out that there wasn't a lot stopping these tests from being in daemon_test, so I went ahead and did that too.

**Review by pedronis — APPROVED** [review]
thank you

**Review by bboozzoo — APPROVED** [review]
LGTM. Thank you

**Pull request merged**